File: phaseGateOne/BankeBank/MenstrualCalculator.java
import java.time.LocalDate;
import java.time.Period;
import java.util.ArrayList;
import java.util.Scanner;

public class MenstrualCalculator {
    private LocalDate flowStartDate;
    private LocalDate flowEndDate;
    private int cycleLength = 28;
    
    public MenstrualCalculator(LocalDate flowStartDate, LocalDate flowEndDate) {
      this.flowStartDate = flowStartDate;
      this.flowEndDate = flowEndDate;
    }
    
    public void setCycleLength(int cycleLength) {
      this.cycleLength = cycleLength;
    }

   
    public int calculateFlowDuration() {
      return Period.between(flowStartDate, flowEndDate).getDays() + 1;
    }

    
    public LocalDate calculateNextPeriodDate() {
      return flowStartDate.plusDays(cycleLength);
    }
    
    public LocalDate calculateOvulationDate() {
      return calculateNextPeriodDate().minusDays(14);
    }
    
    public ArrayList<LocalDate> calculateFertileWindow() {
        ArrayList<LocalDate> fertileWindow = new ArrayList<>();
        LocalDate ovulationDate = calculateOvulationDate();
        
        for (int i = 5; i >= 0; i--) {
            fertileWindow.add(ovulationDate.minusDays(i));
        }
        
        fertileWindow.add(ovulationDate.plusDays(1));
        
        return fertileWindow;
    }
    
    public ArrayList<LocalDate> calculateSafePeriod() {
        ArrayList<LocalDate> safePeriod = new ArrayList<>();
        LocalDate fertileStart = calculateOvulationDate().minusDays(5);
        LocalDate fertileEnd = calculateOvulationDate().plusDays(1);
        
        
        LocalDate currentDate = flowEndDate.plusDays(1);
        while (currentDate.isBefore(fertileStart)) {
            safePeriod.add(currentDate);
            currentDate = currentDate.plusDays(1);
        }
        
        
        currentDate = fertileEnd.plusDays(1);
        LocalDate nextPeriod = calculateNextPeriodDate();
        while (currentDate.isBefore(nextPeriod)) {
            safePeriod.add(currentDate);
            currentDate = currentDate.plusDays(1);
        }
        
        return safePeriod;
    }
    public ArrayList<LocalDate> calculateSafePeriod() {
        ArrayList<LocalDate> safePeriod = new ArrayList<>();
        LocalDate fertileStart = calculateOvulationDate().minusDays(5);
        LocalDate fertileEnd = calculateOvulationDate().plusDays(1);
        

        LocalDate currentDate = flowEndDate.plusDays(1);
        while (currentDate.isBefore(fertileStart)) {
            safePeriod.add(currentDate);
            currentDate = currentDate.plusDays(1);
        }
        

        currentDate = fertileEnd.plusDays(1);
        LocalDate nextPeriod = calculateNextPeriodDate();
        while (currentDate.isBefore(nextPeriod)) {
            safePeriod.add(currentDate);
            currentDate = currentDate.plusDays(1);
        }
        
        return safePeriod;
    }
    
    public ArrayList<LocalDate> calculatePreMenstrualSyndromeDays() {
        ArrayList<LocalDate> pmsDays = new ArrayList<>();
        LocalDate nextPeriod = calculateNextPeriodDate();
        
        for (int i = 7; i > 0; i--) {
            pmsDays.add(nextPeriod.minusDays(i));
        }
        
        return pmsDays;
    }
    
    public String getFertilityStatus(LocalDate date) {
        if (date.isEqual(flowStartDate) || 
            (date.isAfter(flowStartDate) && date.isBefore(flowEndDate.plusDays(1)))) {
            return "Menstrual Flow";
        }
        
        if (calculateFertileWindow().contains(date)) {
            if (date.isEqual(calculateOvulationDate())) {
                return "Ovulation Day - Highest Fertility";
            }
            return "Fertile Window - High Fertility";
        }
        
        if (calculateSafePeriod().contains(date)) {
            return "Safe Period - Low Fertility";
        }
        
        if (calculatePMSDays().contains(date)) {
            return "PMS Window";
        }
        
        return "Regular Day";
    }
    
    public static void main(String[] args) {
        
      LocalDate flowStart = LocalDate.now();
      LocalDate flowEnd = flowStart.plusDays(5);
      
      MenstrualCalculator calculator = new MenstrualCalculator(flowStart, flowEnd);
      
      System.out.println("Flow Duration: " + " days");
      System.out.println("Next Period: " );
      System.out.println("Ovulation Date: " );
      
      System.out.println("\nFertile Window:");
      
      
      System.out.println("\nSafe Period Days:");
      
      
      System.out.println("\nPre-Menstrual Syndrome Period:");
    }
}

File: phaseGateOne/python/menstrualcalculator.py
from datetime import date, timedelta


class MenstrualCalculator:
    def __init__(self, flow_start_date: date, flow_end_date: date):
        self.flow_start_date = flow_start_date
        self.flow_end_date = flow_end_date
        self.cycle_length = 28
    
    def set_cycle_length(self, cycle_length: int):
        self.cycle_length = cycle_length
    
    def calculate_flow_duration(self) -> int:
        return (self.flow_end_date - self.flow_start_date).days + 1
    
    def calculate_next_period_date(self):
        return self.flow_start_date + timedelta(days=self.cycle_length)
    
    def calculate_ovulation_date(self):
        return self.calculate_next_period_date() - timedelta(days=14)
    
    def calculate_fertile_window(self):
        fertile_window = []
        ovulation_date = self.calculate_ovulation_date()
        
        
        for i in range(3, -1, -1):
            fertile_window.append(ovulation_date - timedelta(days=i))
        
        
        fertile_window.append(ovulation_date + timedelta(days=1))
        
        return fertile_window
    
    def calculate_safe_period(self) -> :
        safe_period = []
        fertile_start = self.calculate_ovulation_date() - timedelta(days=5)
        fertile_end = self.calculate_ovulation_date() + timedelta(days=1)
        
        current_date = self.flow_end_date + timedelta(days=1)
        while current_date < fertile_start:
            safe_period.append(current_date)
            current_date += timedelta(days=1)
        

        current_date = fertile_end + timedelta(days=2)
        next_period = self.calculate_next_period_date() - timedelta(days=3)
        while current_date < next_period:
            safe_period.append(current_date)
            current_date += timedelta(days=1)
        
        return safe_period
    
    def calculate_pre_menstrual_syndrome_days(self)cahr:
        pms_days = []
        next_period = self.calculate_next_period_date()
        
        for i in range(7, 0, -1):
            pms_days.append(next_period - timedelta(days=i))
        
        return pms_days
    
    def get_fertility_status(self, check_date: date) -> str:

        if (check_date == self.flow_start_date or 
            (check_date > self.flow_start_date and check_date < self.flow_end_date + timedelta(days=1))):
            return "Menstrual Flow"
        

        if check_date in self.calculate_fertile_window():
            if check_date == self.calculate_ovulation_date():
                return "Ovulation Day - Highest Fertility"
            return "Fertile Window - High Fertility"
        

        if check_date in self.calculate_safe_period():
            return "Safe Period - Low Fertility"
        

        if check_date in self.calculate_pre_menstrual_syndrome_days():
            return "Pre-Menstrual Syndrome Window"
        
        return "Regular Day"

def main():
    print("Enter Last period start date\n")
    year = int(input("Enter year: "))
    month = int(input("Enter month: "))
    day = int(input("Enter day: "))
    
    flow_start = date(year, month, day)
    flow_end = flow_start + timedelta(days=5)
    
    calculator = MenstrualCalculator(flow_start, flow_end)
    
    print(f"Flow Duration: {calculator.calculate_flow_duration()} days")
    print(f"Next Period: {calculator.calculate_next_period_date()}")
    print(f"Ovulation Date: {calculator.calculate_ovulation_date()}")
    
    print("\nFertile Window:\n")
    for d in calculator.calculate_fertile_window():
        print(f"{d} -> {calculator.get_fertility_status(d)}")
    
    print("\nSafe Period Days:\n")
    for d in calculator.calculate_safe_period():
        print(d)
    
    print("\nPre-Menstrual Syndrome Period:\n")
    for d in calculator.calculate_pre_menstrual_syndrome_days():
        print(d)

if __name__ == "__main__":
    main()
